mlflow/tracking/client.py

```diff
@@ -1010,6 +1010,62 @@ def log_dict(self, run_id, dictionary, artifact_file):
                 else:
                     json.dump(dictionary, f, indent=2)
 
+    def log_figure(self, run_id, figure, artifact_file):
+        """
+        Log a figure as an artifact. The following figure objects are supported:
+
+        - `matplotlib.figure.Figure`_
+        - `plotly.graph_objects.Figure`_
+
+        .. _matplotlib.figure.Figure:
+            https://matplotlib.org/api/_as_gen/matplotlib.figure.Figure.html
+
+        .. _plotly.graph_objects.Figure:
+            https://plotly.com/python-api-reference/generated/plotly.graph_objects.Figure.html
+
+        :param run_id: String ID of the run.
+        :param figure: Figure to log.
+        :param artifact_file: The run-relative artifact file path in posixpath format to which
+                              the figure is saved (e.g. "dir/file.png").
+
+        .. code-block:: python
+            :caption: Matplotlib Example
+
+            import mlflow
+            import matplotlib.pyplot as plt
+
+            fig, ax = plt.subplots()
+            ax.plot([0, 1], [2, 3])
+
+            run = client.create_run(experiment_id="0")
+            client.log_figure(run.info.run_id, fig, "figure.png")
+
+        .. code-block:: python
+            :caption: Plotly Example
+
+            import mlflow
+            from plotly import graph_objects as go
+
+            fig = go.Figure(go.Scatter(x=[0, 1], y=[2, 3]))
+
+            run = client.create_run(experiment_id="0")
+            client.log_figure(run.info.run_id, fig, "figure.html")
+        """
+        try:
+            import plotly
+            from matplotlib import pyplot as plt
+            from plotly import graph_objects as go
+        except ImportError as ie:
+            print("Failed to import required libraries: {}".format(ie))
+
+        with self._log_artifact_helper(run_id, artifact_file) as tmp_path:
+            if isinstance(figure, plt.Figure):
+                figure.savefig(tmp_path)
+            elif isinstance(figure, go.Figure):
+                plotly.offline.plot(
+                    figure, filename=tmp_path, include_plotlyjs="cdn", auto_open=False
+                )
```

Comment:
We probably want to throw here if the object isn't a supported figure. Also one question, if we raise an exception on an unsupported figure type here, will we just end up logging an empty artifact file? Not sure how `self._log_artifact_helper` will behave in that case

Reply:
In that case, the process would stop at the line where the exception is thrown.

```python
import contextlib

@contextlib.contextmanager
def foo():
    yield
    print("yes")


with foo():
    raise Exception
```

output:

```
Traceback (most recent call last):
  File "a.py", line 11, in <module>
    raise Exception
Exception
```

Reply:
We could validat the object **before** `self._log_artifact_helper`.